testdriver-vendor.js uses an incorrect position for an action with an specified element origin

rr-codes opened the pull request:
#### 701a3c6a95ff6704100315bc68530ea9afdb4987
<pre>
testdriver-vendor.js uses an incorrect position for an action with an specified element origin
<a href="https://bugs.webkit.org/show_bug.cgi?id=262220">https://bugs.webkit.org/show_bug.cgi?id=262220</a>
<a href="https://rdar.apple.com/116145770">rdar://116145770</a>

Reviewed by Tim Nguyen.

There are two main issues when trying to acquire a position for an action with a specified element origin:

1. Currently, the top-left position is used, whereas the spec (<a href="https://www.w3.org/TR/webdriver2/#processing-actions)">https://www.w3.org/TR/webdriver2/#processing-actions)</a> says to use the center.

2. The check `x instanceof Y` is not correct for all cases; specifically, if `x` exists in a different
&quot;realm&quot; than Y, the check will fail (per <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof#instanceof_and_multiple_realms).">https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof#instanceof_and_multiple_realms).</a>
This happens in `testdriver-vendor.js` when an element is checked if it is an `instanceof` `Element`,
which will fail when the element belongs to a sub-frame.

Fix (1) by just computing the center instead, and fix (2) by using the `Element` from the element&apos;s realm instead of the default one.

* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
(async dispatchMouseActions):

Canonical link: <a href="https://commits.webkit.org/282257@main">https://commits.webkit.org/282257@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/647d3356fa3818bb7cd7bf542cf2f74a7a3c6536

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62570 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41925 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15165 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66554 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13122 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49612 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13458 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50413 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9049 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65639 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38961 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54192 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31155 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35683 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11527 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12050 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57269 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11843 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68283 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6516 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11530 "Found 1 new test failure: workers/worker-user-gesture.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57787 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6545 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54239 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57995 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5437 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9425 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37725 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38810 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39907 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38554 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
```diff
@@ -33,11 +33,11 @@ async function dispatchMouseActions(actions, pointerType)
         case "pointerMove":
             const origin = { x: 0, y: 0 };
-            if (action.origin instanceof Element) {
+            if (action.origin instanceof action.origin?.ownerDocument.defaultView?.Element) {
```
```suggestion
            if (action.origin instanceof action.origin?.ownerDocument?.defaultView?.Element) {
```

The origin can be a string, so we should assume ownerDocument can be null too.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
```diff
@@ -97,10 +97,10 @@ async function dispatchTouchActions(actions, options = { insertPauseAfterPointer
         case "pointerMove":
             touch.phase = "moved";
-            if (action.origin instanceof Element) {
+            if (action.origin instanceof action.origin?.ownerDocument.defaultView?.Element) {
```
ditto

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
```diff
@@ -231,11 +231,11 @@ async function dispatchWheelActions(actions)
             if (duration === undefined)
                 duration = computeTickDuration(actions, "wheel");
-            if (origin instanceof Element) {
+            if (origin instanceof origin?.ownerDocument.defaultView?.Element) {
```
ditto

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
```diff
@@ -33,11 +33,11 @@ async function dispatchMouseActions(actions, pointerType)
         case "pointerMove":
             const origin = { x: 0, y: 0 };
-            if (action.origin instanceof Element) {
+            if (action.origin?.ownerDocument?.defaultView && action.origin instanceof action.origin.ownerDocument.defaultView.Element) {
```
```suggestion
            if (action.origin instanceof action.origin?.ownerDocument?.defaultView?.Element) {
```
Doesn't this work?

rr-codes replied:
I had this originally, but It did not work because in the case where it is null, a JS exception is thrown because the right hand side of `instanceof` must be a non null object

nt1m replied:
Maybe store the window object in a variable to avoid duplicatation?

rr-codes replied:
will do; I also need to figure out what to do about the newly failing tests 😔

rr-codes committed:
testdriver-vendor.js uses an incorrect position for an action with an specified element origin
https://bugs.webkit.org/show_bug.cgi?id=262220
rdar://116145770

Reviewed by Tim Nguyen.

There are two main issues when trying to acquire a position for an action with a specified element origin:

1. Currently, the top-left position is used, whereas the spec (https://www.w3.org/TR/webdriver2/#processing-actions) says to use the center.

2. The check `x instanceof Y` is not correct for all cases; specifically, if `x` exists in a different
"realm" than Y, the check will fail (per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof#instanceof_and_multiple_realms).
This happens in `testdriver-vendor.js` when an element is checked if it is an `instanceof` `Element`,
which will fail when the element belongs to a sub-frame.

Fix (1) by just computing the center instead, and fix (2) by using the `Element` from the element's realm instead of the default one.

* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
(async dispatchMouseActions):

Canonical link: https://commits.webkit.org/282257@main
